### PR TITLE
Add material design styles to baseframe

### LINKED ISSUE #104: Move jinja2 templates to .html.jinja2 extensions
Following from #103, we need to rename all templates to `.html.jinja2` extensions.

However, several templates including `baseframe.html`, `layout.html`, `baseframe/forms.html` and others are referred to from client apps, so renaming them will break clients. We have three options:
1. Don't rename them.
2. Rename but place symlinks at the old names (will possibly break on Windows-based dev environments, apart from creating chaos on Dropbox and other cloud-synced developer filesystem environments).
3. Create shadow templates at the old names that simply consist of `{% extends "<template>.html.jinja2" -%}`. This will work for layout templates but not macro templates.

Approach 2 seems the best at this time.

### PR DESCRIPTION
**Pull request by vidya-ram**
1. Add material design styles to baseframe
    a) Add mui.css- material design css framework(https://github.com/muicss/mui) to baseframe
    b)  Add reflex.css(https://github.com/leejordan/reflex) for responsive flexbox grid support
    c)  Add material design components chip, card, alert, thumbnail to baseframe
2. Maintain bootstrap3 and mui templates in separate folders.
3. Fixes #104

### CODE REVIEW

File: baseframe/forms/auto.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ class ConfirmDeleteForm(Form):
     cancel = SubmitField(__(u"Cancel"))
 
 
-def render_form(form, title, message='', formid='form', submit=__(u"Submit"), cancel_url=None, ajax=False):
+def render_form(form, title, message='', formid='form', submit=__(u"Submit"), cancel_url=None, ajax=False, theme=None):
```
Comment (jace) [inline baseframe/forms/auto.py]: Instead of explicitly passing a theme, could we put this in the config somewhere?

1. In the app's `settings.py`
2. In Baseframe's `init_app`, which is typically called along with a `requires` clause, making maintenance easier. Baseframe can then save it in `app.config` and consult that on each request.
3. In the asset system itself, so that whenever a particular asset is requested, it sets a request-level flag that this function can consult.

My sense is the second method is the most straightforward and will cause the least future maintenance hassle.

File: baseframe/templates/baseframe/ajaxform.html [hunk 2]
```diff
@@ -1,4 +1,8 @@
-{% from "baseframe/forms.html" import renderform, ajaxform, widget_ext_scripts, widgetscripts %}
+{% if theme == "material" %}
+	{% from "baseframe/forms-material.html" import renderform, ajaxform, widget_ext_scripts, widgetscripts %}
+{%- else %}
+    {% from "baseframe/forms.html" import renderform, ajaxform, widget_ext_scripts, widgetscripts %}
```
Comment (jace) [inline baseframe/templates/baseframe/ajaxform.html]: I'd like to have a clean structure to these. Maybe `baseframe/mui/forms.html` and `baseframe/bootstrap3/forms.html`. There is also #104 that requires these templates to have `.html.jinja2` extensions. We might as well do that now, so make this `baseframe/mui/forms.html.jinja2`.

[Since a lot of other repositories refer to `baseframe/forms.html`](https://github.com/search?q=org%3Ahasgeek+from+%22baseframe%2Fforms.html%22&type=Code) we can help transition them by retaining a `forms.html` that simply imports everything from `baseframe/bootstrap3/forms.html`.

File: baseframe/templates/baseframe/components.html [hunk 3]
```diff
@@ -11,6 +11,19 @@
   {% endwith %}
 {%- endmacro %}
 
+{% macro flash_messages_material() -%}
```
Comment (jace) [inline baseframe/templates/baseframe/components.html]: Let's keep the existing function but add an optional `theme` parameter that defaults to `bootstrap3`.

File: baseframe/templates/baseframe/components.html [hunk 4]
```diff
@@ -11,6 +11,19 @@
...
+  {% with messages = get_flashed_messages(with_categories=true) %}
+    {% if messages %}
+      {% for category, message in messages %}
```
Comment (jace) [inline baseframe/templates/baseframe/components.html]: All uses of `{%` should be `{%-` to remove excess whitespace in the rendered template (the original function needs fixing too).

File: baseframe/templates/baseframe/forms-material.html [hunk 5]
```diff
@@ -0,0 +1,334 @@
+{% macro rendererrors(field) -%}
+    {%- if field.errors -%}
+      {%- for error in field.errors %}
+        <div><p class="mui-form--error">{{ error }}</p></div>
+      {%- endfor %}
+    {%- endif %}
+{%- endmacro %}
+
+{% macro renderfield(field, css_class="", widget_css_class="", sidetext=none, tabindex=none, autofocus=false, nolabel=false, style='horiz', rows=none) -%}
```
Comment (jace) [inline baseframe/templates/baseframe/forms-material.html]: This macro may need to be rewritten to remove all the accumulated cruft, but we can do this in a separate PR.

File: baseframe/templates/baseframe/forms-material.html [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+          $(this).addClass('submit-disabled').prop('disabled', true);
+          $('#{{ formid }}').find(".loading").removeClass('hidden');
+          this.form.submit();
+        });
+      });
+    </script>
+  {%- endif %}
+{%- endmacro %}
+
+{% macro tinymce3_editor(field, script=true) -%}
```
Comment (jace) [inline baseframe/templates/baseframe/forms-material.html]: We can drop support for TinyMCE3 in Baseframe-MUI.

File: baseframe/assets.py [hunk 7]
```diff
@@ -300,6 +306,15 @@
     'select2-baseframe.js',
     'getdevicepixelratio.js'
 ]}
+assets['extra-material.js'][Version('0.0.0')] = {'requires': [
+    'jquery-modal.js',
+    'jquery.form.js',
+    'jquery.autosize.js',
+    'jquery.timepicker.js',
+    'select2-material.js',
+    'jquery.tinymce.js>=3.0.0,<4.0.0',
```
Comment (jace) [inline baseframe/assets.py]: TinyMCE should not be bundled with extras unless the app explicitly wants it. TinyMCE3 is also obsolete and only present because of unmaintained sites like [Kharcha](https://expenses.hasgeek.com).

File: baseframe/forms/auto.py (hunk 1 above)
Comment (vidya-ram) [reply baseframe/forms/auto.py]: Passing `theme` in `init_app` and storing it in `app.config`

File: baseframe/templates/baseframe/ajaxform.html (hunk 2 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/ajaxform.html]: Moved bootstrap3 and mui templates to separate folders.
Changed templates extensions to to .html.jinja2

File: baseframe/templates/baseframe/components.html (hunk 3 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/components.html]: Fixed.  Created a new components template for mui.

File: baseframe/templates/baseframe/components.html (hunk 4 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/components.html]: Fixed

File: baseframe/templates/baseframe/forms-material.html (hunk 6 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/forms-material.html]: Removed TinyMCE3 in Baseframe-MUI template
Updated hasgeektv to use TinyMCE4

File: baseframe/assets.py (hunk 7 above)
Comment (vidya-ram) [reply baseframe/assets.py]: Fixed

File: baseframe/__init__.py [hunk 8]
```diff
@@ -196,6 +196,8 @@ def init_app(self, app, requires=[], ext_requires=[], bundle_js=None, bundle_css
 
         app.config['tz'] = timezone(app.config.get('TIMEZONE', 'UTC'))
 
+        app.config['theme'] = theme
```
Comment (jace) [inline baseframe/__init__.py]: Let's validate the theme before doing this, by having a list of recognised theme names at the top of this file. Basically:

```python
THEME_FILES = {
    'bootstrap3': {'autoform.html.jinja2': 'baseframe/bootstrap3/autoform.html.jinja2', …},
    'mui': {'autoform.html.jinja2': 'baseframe/mui/autoform.html.jinja2', …}
    }
…
if theme not in THEME_FILES:
    raise ValueError("Unrecognised theme: %s" % theme)
```

File: baseframe/forms/auto.py [hunk 9]
```diff
@@ -26,25 +26,37 @@ def render_form(form, title, message='', formid='form', submit=__(u"Submit"), ca
     else:
         code = 200
     if request.is_xhr and ajax:
-        return make_response(render_template('baseframe/ajaxform.html', form=form, title=title,
+        template = 'baseframe/bootstrap3/ajaxform.html.jinja2'
+        if current_app.config['theme'] == "mui":
```
Comment (jace) [inline baseframe/forms/auto.py]: `'mui'` in single quotes everywhere as it's an identifier, not a UI string.

File: baseframe/forms/auto.py [hunk 10]
```diff
@@ -26,25 +26,37 @@ def render_form(form, title, message='', formid='form', submit=__(u"Submit"), ca
...
+            template = 'baseframe/mui/ajaxform.html.jinja2'
```
Comment (jace) [inline baseframe/forms/auto.py]: Make this entire `if` condition simpler by doing this:
```python
template = THEME_FILES[current_app.config['theme']]['ajaxform.html.jinja2']
```

File: baseframe/__init__.py (hunk 8 above)
Comment (vidya-ram) [reply baseframe/__init__.py]: Fixed

File: baseframe/forms/auto.py (hunk 9 above)
Comment (vidya-ram) [reply baseframe/forms/auto.py]: Removed.

File: baseframe/forms/auto.py (hunk 10 above)
Comment (vidya-ram) [reply baseframe/forms/auto.py]: Fixed

File: baseframe/__init__.py [hunk 11]
```diff
@@ -43,6 +43,15 @@
...
 }
 
+THEME_FILES = {
+    'bootstrap3': {'ajaxform.html.jinja2': 'baseframe/bootstrap3/ajaxform.html.jinja2', 'autoform.html.jinja2': 'baseframe/bootstrap3/autoform.html.jinja2',
+        'delete.html.jinja2': 'baseframe/bootstrap3/delete.html.jinja2', 'message.html.jinja2': 'baseframe/bootstrap3/message.html.jinja2',
+        'redirect.html.jinja2': 'baseframe/bootstrap3/redirect.html.jinja2'},
+    'mui': {'ajaxform.html.jinja2': 'baseframe/mui/ajaxform.html.jinja2', 'autoform.html.jinja2': 'baseframe/mui/autoform.html.jinja2',
+        'delete.html.jinja2': 'baseframe/mui/delete.html.jinja2', 'message.html.jinja2': 'baseframe/mui/message.html.jinja2',
+        'redirect.html.jinja2': 'baseframe/mui/redirect.html.jinja2'}
+}
```
Comment (jace) [inline baseframe/__init__.py]: A note on indentation: one file per line, and the closing `}` should be indented (it violates PEP8 but is part of our style guide).

File: baseframe/forms/fields.py [hunk 12]
```diff
@@ -203,7 +203,9 @@ def __init__(self,
         tinymce_options.setdefault('plugins', "autolink autoresize link lists paste searchreplace")
         tinymce_options.setdefault('toolbar', "bold italic | bullist numlist | link unlink | searchreplace undo redo")
         tinymce_options.setdefault('width', "100%")
-        tinymce_options.setdefault('height', "400")
+        tinymce_options.setdefault('height', "200")
+        tinymce_options.setdefault('autoresize_min_height', "159")
+        tinymce_options.setdefault('autoresize_max_height', "200")
```
Comment (jace) [inline baseframe/forms/fields.py]: All of the right-hand-side values should be in single quotes. This is pre-existing code, but might as well fix it now that this file is being modified.

File: baseframe/templates/baseframe/bootstrap3/autoform.html.jinja2 [hunk 13]
```diff
@@ -1,4 +1,4 @@
-{% extends "baseframe/bootstrap3/formlayout.html.jinja2" %}
+{% extends "formlayout.html" %}
```
Comment (jace) [inline baseframe/templates/baseframe/bootstrap3/autoform.html.jinja2]: This should not refer to `formlayout.html` since that file is deprecated.

File: baseframe/templates/baseframe/bootstrap3/delete.html.jinja2 [hunk 14]
```diff
@@ -1,4 +1,4 @@
-{% extends "baseframe/bootstrap3/formlayout.html.jinja2" %}
+{% extends "formlayout.html" %}
```
Comment (jace) [inline baseframe/templates/baseframe/bootstrap3/delete.html.jinja2]: Any reason for reverting to `formlayout.html`?

File: baseframe/templates/baseframe/bootstrap3/autoform.html.jinja2 (hunk 13 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/bootstrap3/autoform.html.jinja2]: Hasjob has its own formlayout.html

File: baseframe/templates/baseframe/bootstrap3/delete.html.jinja2 (hunk 14 above)
Comment (vidya-ram) [reply baseframe/templates/baseframe/bootstrap3/delete.html.jinja2]: Since Hasjob has its own  formlayout.html

File: baseframe/__init__.py (hunk 11 above)
Comment (vidya-ram) [reply baseframe/__init__.py]: Fixed

File: baseframe/forms/fields.py (hunk 12 above)
Comment (vidya-ram) [reply baseframe/forms/fields.py]: Fixed
